### ISSUE TITLE
Build on NetBSD 9.1 broken

### ISSUE THREAD

**Issue by edgarpettijohn** [OpenSMTPD/OpenSMTPD#1114]
The configure process picks up the local strtonum() and reallocarray(). However, to use them you have to define _OPENBSD_SOURCE. Not sure how or where the best place to do that is.  Unfortunantly neither of these causes the build to error out so upon testing the first time smtpd would die upon receiving the first line after DATA with nothing in the log.

Undefining the HAVE_STRTONUM and HAVE_REALLOCARRAY in config.h allows to build and run properly.

**Comment by poolpOrg** [comment]
It's unfortunate that this gets reported after the release, it was broken in the release candidate :-(

I'll commit a real fix shortly, maybe tomorrow, but it's unlikely I'll make another release before end of next week to let other people report if they have issues with the release.

**Comment by edgarpettijohn** [comment]
Sorry. Forgot the machine existed until the power was inadvertently disrupted. Decided to upgrade it and then finally got around to updating smtpd yesterday.

**Comment by poolpOrg** [comment]
Don't apologize, it happens, but unluckily it happens to me at every release 😅

**Comment by ngortheone** [comment]
Hey @edgarpettijohn 
Can you take a look at https://github.com/OpenSMTPD/OpenSMTPD/issues/1120 and see if you can help? 
I don't have enough expertise with NetBSD so I am looking for someone interested and with NetBSD experience.

**Comment by edgarpettijohn** [comment]
I'll take a look, but no promises. I can't stand qemu.

**Comment by edgarpettijohn** [comment]
I've made a few changes to the genimg script. Testing it now.

**Comment by mtbc** [comment]
Building from this repository on NetBSD 9.1 I've not gotten any further than,
```
undefined reference to `IN6_IS_ADDR_MC_INTFACELOCAL'
```
but I'm happy to take that to a different issue and I may be missing some important arguments to `--configure`.

I see that git://wip.pkgsrc.org/pkgsrc-wip.git includes an opensmtpd based on 6.8.0p2, with patches. It builds and kinds of runs for me, if I fix their `--enable-table-db` to `--with-table-db` then the db driver works too, but then I observe exactly the above,

> ... the first time smtpd would die upon receiving the first line after DATA with nothing in the log.

The log simply notes, `smtpd: process pony socket closed`.

From either starting point (this upstream or NetBSD wip), I'm happy to try things and report back if that's of any use, I don't know enough to fix much myself I'm afraid.

**Comment by mtbc** [comment]
I should add, from the pkgsrc wip version that I can actually build, I see the same symptom even with,
```c
#undef HAVE_REALLOCARRAY
#undef HAVE_STRTONUM
```
(I set that in between configuring and doing the rest of the build.)

Nevertheless, I'm curious as to how you narrowed your issue to this, in case I can do likewise with my issue.

**Comment by edgarpettijohn** [comment]
On Sun, May 23, 2021 at 04:18:10PM -0700, Mark Carroll wrote:
> I should add, from the pkgsrc wip version that I can actually build, I see the same symptom even with,
> ```c
> #undef HAVE_REALLOCARRAY
> #undef HAVE_STRTONUM
> ```
> 
> Nevertheless, I'm curious as to how you narrowed your issue to this, in case I can do likewise with my issue.
>

I don't recall at the moment. My NBSD machine has died since then. More
than likely I built it with `-g' and ran it with gdb to see what was
going on. I'm trying to spin up a vm and see what I get.

Edgar

**Comment by edgarpettijohn** [comment]
On Sun, May 23, 2021 at 04:18:10PM -0700, Mark Carroll wrote:
> I should add, from the pkgsrc wip version that I can actually build, I see the same symptom even with,
> ```c
> #undef HAVE_REALLOCARRAY
> #undef HAVE_STRTONUM
> ```
> 
> Nevertheless, I'm curious as to how you narrowed your issue to this, in case I can do likewise with my issue.
>

Just finished ./configure && make to see a bunch of warning implicite
declaration of function strtonum... So thats where I got started on that
path.

Edgar

**Comment by mtbc** [comment]
Aha, good point! I get those and also for function 'err' in `smtpd/util.c`. After configuring, if I add,
```c
#undef HAVE_ERR
#undef HAVE_ERRC
#undef HAVE_ERRX
#undef HAVE_ERR_H
```
I still have the same MTA failure, alas, also still warnings about implicit 'err', so I may have to pull out `gdb` before we're done, or try to make sense of the other warnings.

**Comment by mtbc** [comment]
```c
#undef HAVE_CLOSEFROM_INT
```
(together with the above) gets rid of that last implicit warning but doesn't help overall.

**Comment by edgarpettijohn** [comment]
On May 23, 2021 9:16 PM, Mark Carroll ***@***.***> wrote:
#undef HAVE_CLOSEFROM_INT
(together with the above) gets rid of that last implicit warning but doesn't help overall.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
That seems odd. I recommend getting the latest stable portable tarball from opensmtpd.org and start from there so as to simplify matters.Edgar

**Comment by mtbc** [comment]
Given the `IN6_IS_ADDR_MC_INTFACELOCAL` issue in https://github.com/OpenSMTPD/OpenSMTPD/issues/1114#issuecomment-846619711, are you adding any `configure` arguments I should also try? It's not as if I need IPv6 right now anyway. Though the pkgsrc wip version doesn't fiddle with anything from 6.8.0p2 that looks relevant.

**Comment by edgarpettijohn** [comment]
nope. Just a plain ole `./configure && make'. Granted I did start off from NetBSD 9.2 as I didn't feel like finding 9.1 for the throwaway vm. I suspect something is very broken with your build environment. After completing the build and install. It did die due to something with file permissions for smtpq, most likely /var/spool/whatever probably didn't exist or the permissions were wrong, but I didn't look into it further.

**Comment by omar-polo** [comment]
Hello,

I haven't seen this issue before, otherwise I would have mentioned that a few portability fixes were included in 7.3.0.  After a report on the mailing list I fired up a NetBSD vm and fixed the build there.  I don't use NetBSD daily, so I only verified that it built there and did some brief runtime testing.

Is this still an issue on NetBSD?  

Unfortunately the current configure script is quite fragile and can pick up functions when it should not, that's the main motivation behind #1222 that I hope to have ready in time for 7.4, but even the current version should build fine on NetBSD.
